src/sql/src/plan/explain/text.rs

```diff
@@ -332,19 +333,70 @@ impl fmt::Display for HirScalarExpr {
             }
             Windowing(expr) => {
-                match &expr.func {
-                    WindowExprType::Scalar(scalar) => {
-                        write!(f, "{}()", scalar.clone().into_expr())?
+                // First, print
+                // - the window function name
+                // - the arguments.
+                // Also, dig out some info from the `func`.
+                let (column_orders, ignore_nulls, window_frame) = match &expr.func {
+                    WindowExprType::Scalar(scalar_window_expr) => {
+                        write!(f, "{}()", scalar_window_expr.func)?;
+                        (&scalar_window_expr.order_by, false, None)
                     }
-                    WindowExprType::Value(scalar) => {
-                        write!(f, "{}({})", scalar.clone().into_expr(), scalar.expr)?
+                    WindowExprType::Value(value_window_expr) => {
+                        write!(f, "{}({})", value_window_expr.func, value_window_expr.args)?;
+                        (
+                            &value_window_expr.order_by,
+                            value_window_expr.ignore_nulls,
+                            Some(&value_window_expr.window_frame),
+                        )
                     }
+                };
+
+                // Reconstruct the ORDER BY (see comment on `WindowExpr.order_by`).
+                // We assume that the `column_order.column`s refer to each of the expressions in
+                // `expr.order_by` in order. This is a consequence of how `plan_function_order_by`
+                // works.
+                assert!(column_orders
+                    .iter()
+                    .enumerate()
+                    .all(|(i, column_order)| i == column_order.column));
+                let order_by = column_orders
+                    .iter()
+                    .zip_eq(expr.order_by.iter())
+                    .map(|(column_order, expr)| {
+                        ColumnOrderWithExpr {
+                            expr: expr.clone(),
+                            desc: column_order.desc,
+                            nulls_last: column_order.nulls_last,
+                        }
+                        // (We can ignore column_order.column because of the above assert.)
+                    })
+                    .collect_vec();
+
+                // Print IGNORE NULLS if present.
+                if ignore_nulls {
+                    write!(f, " ignore nulls")?;
                 }
-                write!(f, " over ({})", separated(", ", expr.partition.iter()))?;
 
-                if !expr.order_by.is_empty() {
-                    write!(f, " order by ({})", separated(", ", expr.order_by.iter()))?;
+                // Print the OVER clause.
+                // This is close to the SQL syntax, but we are adding some [] to make it easier to
+                // read.
+                write!(f, " over (")?;
```

Comment:
Is it OK / consistent with SQL to just print `over (order by [])` in case all components rendered within the `( ... )` are empty?

Reply:
SQL syntax wouldn't allow an empty `ORDER BY`, but it allows completely omitting the `ORDER BY`:
```
row_number() over (order by)
```
is a syntax error, but
```
row_number() over ()
```
is allowed.

In the HIR EXPLAIN, we are going to print the latter as `over (order by [])`. I could have written the explain code to print `over ()`, but I decided against this for 2 reasons:
- Omitting the ORDER BY from an OVER clause is rare, since window functions are all about order. If somebody omits it, then it might be good if an empty ORDER BY jumps out at them in the EXPLAIN.
- And a practical reason was that inside the `over (...)` there can be 3 things (partition by, order by, window frame), and if all 3 of these things are optional, then it takes some fiddling to correctly print spaces between them in all 2 * 2 * 2 cases of which of these are present. On the other hand, if I print the middle one always (the order by), then there is no extra conditional logic needed for inserting spaces.